Enable L7 DNS proxy for nodes

atykhyy opened the pull request:
FQDN-based network policy is extremely useful for egress filtering, but is currently only supported in pods. This means that nodes must either rely on some other filtering mechanism, or resort to periodically updating L3/L4 CIDR policies. However, the code which collects DNS requests and combines them with policy to automatically produce L3/L4 filters is at endpoint level and is almost agnostic with respect to kind of endpoint (pod or node). This PR enables FQDN-based network policy for nodes by adding a bit of BPF code which, when instructed by network policy, redirects node's DNS requests to the same transparent proxy that is used for pods. No changes in iptables rules or routing are required.

FQDN-based network policy for nodes is likely to be a good addition to Cilium, as it would enable egress filtering by FQDNs for the whole cluster without relying on third-party solutions. Also, the same approach may possibly work for other L7 rules.

A sample policy I used for testing is
```yaml
apiVersion: cilium.io/v2
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: hostdns
specs:
- egress:
  - toCIDR:
    - 168.63.129.16/32 # example DNS server IP (whatever is in cilium agent's /etc/resolv.conf)
    toPorts:
    - ports:
      - port: "53"
        protocol: ANY
      rules:
        dns:
        - matchPattern: '*'
  - toFQDNs: # works when host has L7 DNS policy
    - matchName: api.ipify.org
  nodeSelector: {}
```

nathanjsweet commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -215,12 +215,32 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPortProto restore.PortPr
-		if _, exists := ipRule.IPs[destIP]; exists || len(ipRule.IPs) == 0 {
-			if ipRule.regex != nil && ipRule.regex.MatchString(name) {
-				return true
+		if IPs := ipRule.IPs; IPs == nil {
```
This must be `len(IPs) == 0`, as the map can exist, but be empty.

nathanjsweet commented on daemon/cmd/fqdn.go:
```diff
@@ -180,13 +181,20 @@ func (d *Daemon) updateDNSDatapathRules(ctx context.Context) error {
 
-	return e, nil
+	if node.IsNodeIP(endpointAddr) != "" {
+		if e := d.endpointManager.GetHostEndpoint(); e != nil {
```
We do not currently support this behavior in L3/L4. You are creating a feature for L7 that does not exist for other layers. Please remove this from your PR and propose a way to make it work for L3/L4. [This PR](https://github.com/cilium/cilium/pull/27464) created a feature that made nodes selectable. You can use it as a template for how to make host selectable in L3/L4. Please bring this up in the #sig-policy channel on Cilium Slack or attend a weekly community meeting for further assistance.

atykhyy replied:
If I remove this, the whole feature will stop working, because then the DNS proxy will not be able to determine the Cilium endpoint ID corresponding to host's DNS request and will reject it, logging the message "Cannot extract endpoint ID from DNS request" (pkg/fqdn/dnsproxy/proxy.go:966). Host policy `toFQDNs` rules will then have nothing to work with.

> We do not currently support this behavior in L3/L4.

I do not understand what behavior you mean. Cilium already has the concept "this network flow comes from/to a host". Nodes (including local host) have endpoint IDs and security IDs, they can have network policies on egress and ingress. The only purpose of this part of the PR is to patch this knowledge through to DNS proxy. It does not care what those IP addresses are, it does not enable users to specify node addresses, and it does not expose host IP addresses anywhere they weren't visible or usable before. It is purely internal to Cilium agent. I only need to be able to associate a Cilium endpoint ID with DNS requests coming from the local node (host). If there is a way to do this without involving host IP addresses, which I admit is kludgy, I would be happy to do so. Perhaps DNS proxy could distinguish this by network interface the request came on (`cilium_host` vs anything else), although this would require extending miekg/dns to pass through network interface information in addition to address/port. Another alternative is to run DNS proxy on two ports and redirect host's DNS requests to a different proxy port than pods' DNS requests (this would require creating an extra pair of iptables -j TPROXY rules). The host-dedicated `ServeDNS` would not need to look up request's endpoint ID at all, it would just call `d.endpointManager.getHostEndpoint()` directly. This part of the PR does essentially the same thing.

nathanjsweet replied:
Resolved in this discussion on slack:
https://cilium.slack.com/archives/CDLN7836J/p1722360139578739

nathanjsweet reviewed (changes requested):
The inclusion of "host" in your PR as CIDR selectable needs further discussion. I would like to see a lot of this functionality behind the [`--policy-cidr-match-mode` flag](https://github.com/cilium/cilium/pull/27464/files#diff-26a87034ca4fd33b911176cd448b38fb1c636c2085fc87df2e78836c4b2600e9R282https://github.com/cilium/cilium/pull/27464/files#diff-26a87034ca4fd33b911176cd448b38fb1c636c2085fc87df2e78836c4b2600e9R282) as well.

qmonnet commented on pkg/fqdn/restore/restore.go:
```diff
@@ -75,7 +77,70 @@ type IPRules []IPRule
+
+func MustParseRuleIPOrCIDR(s string) (ip RuleIPOrCIDR) {
+	if err := ip.UnmarshalText([]byte(s)); err != nil {
+		panic(fmt.Errorf("bad input '%s': %w", s, err))
```
Can't we return an error to handle this gracefully instead of panic-king here? The error message could also provide more context.

nathanjsweet replied:
Oh yeah. We don't allow panics in Cilium code.

atykhyy replied:
Sure, I fixed this to log a warning.

qmonnet commented on Documentation/security/policy/language.rst:
```diff
@@ -964,8 +964,9 @@ latter rule will have no effect.
-.. note:: Layer 7 rules are not currently supported in `HostPolicies`, i.e.,
-          policies that use :ref:`NodeSelector`.
+.. note:: Only DNS layer 7 rules are currently supported in `HostPolicies`,
+          i.e., policies that use :ref:`NodeSelector`.
```
This reads like DNS layer 7 rules are those using a `NodeSelector`, when what you mean to say (I believe) is that among layer 7 rules, relying on a `NodeSelector`, only DNS policies are supported.

Can you please fix it?

atykhyy replied:
How about this?

> In `HostPolicies`, i.e. policies that use :ref:`NodeSelector`, only DNS layer 7 rules are supported.

qmonnet replied:
Sounds good to me :+1:

atykhyy replied:
Fixed.

qmonnet reviewed (dismissed):
Nice work! With a couple comments below from my side.

~I'd like to see a conclusion to the discussion with @nathanjsweet, I'm not sure why it was marked as “resolved”?~

atykhyy committed:
Support CIDRs in restored DNS rules properly

Restored DNS rules code expected that LookupIPsBySecID returns only IP addresses,
but if an L7 DNS rule has toCIDR selectors configured, it also returns CIDRs,
and after such a DNS rule is restored, it does not match any target DNS servers
and the DNS proxy rejects DNS queries until the endpoint is regenerated.

This commit fixes this issue, replacing string IP addresses in restore.DNSRules
with a wrapped netip.Prefix (preserving JSON serialization of restored rules).

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>

atykhyy committed:
Basic support for looking up host endpoint by IP

The host endpoint currently does not have any IP addresses associated with it.
This prevents DNS proxy from working when source is the host endpoint.
This commit is a minimal change which enables the host endpoint to be looked up
by its IP address without touching any other code.

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>

atykhyy committed:
Enable L7 DNS proxy for nodes

FQDN-based network policy is extremely useful for egress filtering, but is
currently only supported in pods. This means that nodes must either rely on
some other filtering mechanism, or resort to periodically updating L3/L4 CIDR
policies. However, the code which collects DNS requests and combines them with
policy to automatically produce L3/L4 filters is at endpoint level and is
almost agnostic with respect to kind of endpoint (pod or node). This changeset
enables FQDN-based network policy for nodes by adding BPF code which redirects
node's DNS requests to the same transparent proxy that is used for pods.

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>
